### PR TITLE
Use more StringView

### PR DESCRIPTION
**Pull request by rburchell**

### CODE REVIEW

File: AK/StringView.h [hunk 1]
```diff
@@ -28,6 +28,10 @@ class StringView {
         }
     }
     StringView(const AK::String& string);
+    StringView(const AK::StringView& view, int length)
```
Comment (awesomekling) [inline AK/StringView.h]: This strikes me as a weird API for a constructor.
It's more like a "left()" or "substring()" or something, no?

Comment (rburchell) [reply AK/StringView.h]: Yeah, that's definitely true. I'm not actually sure I ended up needing it anyway.

Comment (rburchell) [reply AK/StringView.h]: Okay, I did, but I've replaced it by using the char*, len ctor for now. I think that some left/right/mid sugar would be nice, but I'd rather do that as a separate change on top.

**Comment by awesomekling** [comment]
Hmm. Passing a `StringView` to someone who immediately stores it into a `String` is a pessimization, since that will now incur a full string copy.
I think we should only take `StringView` as an argument when we don't intend to keep the string around.
Or am I missing something obvious here?

**Comment by rburchell** [comment]
Oh, realised I didn't write what I wrote on IRC here also... Guess I forgot to submit.

The thing I was going for with this was simplicity: if we take the stance that we store String, but accept StringView, then we have one simple "interface" that is easy enough to comprehend (and also then cheap to take any type of string-like data).

Yes, in many cases, it will ultimately need to allocate, but that is not universal, and even when it is, it can (potentially) avoid that unless it's really necessary (via overloads on operators for StringView).

You're probably right that it doesn't help with a large number of cases that then need to subsequently allocate, indeed. Shallow copies of String make sense to me. So the idea would be to use RetainPtr on the impl to just bump the refcount when constructing String from StringView that already had a String? That seems reasonable.

**Comment by awesomekling** [comment]
Yeah exactly. `String` is really just a convenience wrapper around `RetainPtr<StringImpl>`, so I was thinking we could have include an optional `StringImpl*` in `StringView` (but not modify the retain count in StringView, just leave it raw.)
Then `String(StringView)` and `String::operator=(StringView)` can check for the presence of this `StringImpl*` and if present, just assign that to self.

**Comment by rburchell** [comment]
StringView shouldn't retain, obviously, but the construction of a String from StringView should, surely?

**Comment by awesomekling** [comment]
Yes exactly. Sorry for being muddy, I’m on my phone :)

**Comment by rburchell** [comment]
Okay, as discussed, StringView -> String will now be "cheap" when possible.

**Comment by awesomekling** [comment]
V nice!

**Pull request merged**